ocaml/testsuite/tests/typing-local/local.ml

```diff
@@ -1815,6 +1815,12 @@ let foo (local_ x) =
   | _ -> ()
 [%%expect{|
 val foo : local_ int -> unit = <fun>
+|}, Principal{|
+Line 3, characters 21-22:
+3 |   | 0 as y -> escape y
+                         ^
+Error: This value escapes its region
+  Hint: This argument cannot be local, because it is an argument in a tail call
```

Comment:
something is wrong here? I suspect this is because the changes to `Tpat_constant`

Reply:
Well, I thought this was OK, because we could say that we don't know the type of `y` principally. But I think that's a bit bogus. I will fix.

Reply:
After looking into this for a while, I think my original behavior -- no mode-crossing in principal mode -- is correct. The reason is that `y`'s type is not known principally. We can observe this in this program:

```ocaml
type _ g =
  | Int : int g

let f (type a) (x : a g) (y : a) =
  match x, y with
  | Int, (3 as x) -> x
  | _ -> assert false
```

If the type of `x` were known principally, then we could infer the return type of `f` as `int`. But we currently get 

```
Error: This expression has type a = int
       but an expression was expected of type 'a 
       This instance of int is ambiguous:
       it would escape the scope of its equation
```

I know this because I fixed the bug, and the behavior here changed. So I favor just keeping the behavior as originally committed. If anything, we could treat the old behavior, allowing mode-crossing, as a bug.

Note that this does not affect code within Jane Street at all, because we do not use principal mode.

Reply:
I agree that we should only cross mode if `is_principal ty = true`. But I'm slightly confused as to what's causing the change in output. 
Do you mean there is a BUG (independent of mode crossing) wrt principality check?

Reply:
There is a design choice that, in `match blah with 0 as x -> ...`, the type of `x` is not known principally. This makes some sense in the context of GADT pattern-matching, where principality means "the user has made their intent unambiguously clear".

Given this design choice -- and our desire to mode-cross only when a type is principally known -- we shouldn't mode-cross `x` either. Before this patch, we did mode-cross `x`, because there was a missing principality check. Now we don't. In light of the overall design here, I think this is an improvement.

Reply:
Thanks for explanation.